### PR TITLE
Feat/send messages

### PR DESCRIPTION
**Pull request by Marcellofabbri**
-TisSyncService extrapolates the table name (e.g. "Post") and id (e.g. postId) we're interested in fetching.
-TisSyncService sends a message via AmazonSQS to **tis-sync** with these details (table name and post id)
TIS21-1036

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -46,6 +46,10 @@ dependencies {
 
   // Required to support PATCH requests.
   implementation "org.apache.httpcomponents:httpclient:4.5.12"
+
+  implementation 'com.amazonaws:amazon-sqs-java-messaging-lib:1.0.4'
+
+  implementation 'org.springframework:spring-context:5.2.8.RELEASE'
```
Comment (Judge40) [inline build.gradle:52]: Change to `"` for consistency

File: src/main/java/uk/nhs/hee/tis/trainee/sync/config/AmazonSQSConfig.java [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package uk.nhs.hee.tis.trainee.sync.config;
+
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.AmazonSQSClientBuilder;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class AmazonSQSConfig {
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/config/AmazonSQSConfig.java:9]: This will get flagged in the style checker for incorrect casing, should be `AmazonSqsConfig`. Same issue with the method name.

See https://google.github.io/styleguide/javaguide.html#s5.3-camel-case

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/MessageSendingService.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  private String queueName;
+
+  /**
+   * A service that sends messages into a queue.
+   * @param amazonSQS An AmazonSqs object.
+   * @param objectMapper A tool to construct a Json string.
+   * @param queueName The name of the queue.
+   */
+  public MessageSendingService(AmazonSQS amazonSQS,
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/MessageSendingService.java:30]: Incorrect casing for `amazonSQS` parameter

Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/MessageSendingService.java:30]: Consider renaming the service to better reflect its purpose, since its not for sending generic messages. Maybe `DataRequestService` or similar.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/MessageSendingService.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    this.amazonSqs = amazonSQS;
+    this.objectMapper = objectMapper;
+    this.queueName = queueName;
+  }
+
+  public void sendMessage(String tableName, String id) throws JsonProcessingException {
+    String messageBody = makeJson(tableName, id);
+
+    String queueUrl = amazonSqs.getQueueUrl(queueName).getQueueUrl();
+    SendMessageRequest send_msg_request = new SendMessageRequest()
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/MessageSendingService.java:42]: `send_msg_request` should be camel cased instead of underscores.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/MessageSendingService.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+                               @Value("${application.aws.sqs.queueName}") String queueName) {
+    this.amazonSqs = amazonSQS;
+    this.objectMapper = objectMapper;
+    this.queueName = queueName;
+  }
+
+  public void sendMessage(String tableName, String id) throws JsonProcessingException {
+    String messageBody = makeJson(tableName, id);
+
+    String queueUrl = amazonSqs.getQueueUrl(queueName).getQueueUrl();
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/MessageSendingService.java:41]: Maybe move this to the constructor and have a `queueUrl` field instead of `queueName`.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java [hunk 6]
```diff
@@ -52,14 +70,29 @@ public void syncRecord(Record record) {
...
+      case INSERT:
+      case UPDATE:
+        enrichOrRequestPost(record);
+        break;
+      case DELETE:
+    }
   }
 
-  public Set<Placement> findByPostId(String postId) {
-    return repository.findByPostId(postId);
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java:58]: Looks like we lost this method in the merge/rebase

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java [hunk 7]
```diff
@@ -52,14 +70,29 @@ public void syncRecord(Record record) {
...
+    if (fetchedPost.isPresent()) {
+      placementEnricherFacade.enrich((Post)record);
+    } else {
+      postSyncService.request(postId);
+    }
   }
 
-  public void request(String id) {
-    // TODO: Implement.
-    throw new UnsupportedOperationException();
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java:63]: Also lost this one.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java [hunk 8]
```diff
@@ -52,14 +70,29 @@ public void syncRecord(Record record) {
...
-    return repository.findByPostId(postId);
+  public void enrichOrRequestPost(Record record) {
+    String postId = record.getData().get("postId");
+    Optional<Post> fetchedPost = postSyncService.findById(postId);
+
+    if (fetchedPost.isPresent()) {
+      placementEnricherFacade.enrich((Post)record);
+    } else {
+      postSyncService.request(postId);
+    }
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java:92]: We're doing too much in this service, the enricher will handle everything for us. Just need to implement the `request` method.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/PostSyncService.java [hunk 9]
```diff
@@ -33,10 +37,16 @@
 @Service("tcs-Post")
 public class PostSyncService implements SyncService {
 
+  private static final Logger LOG = LoggerFactory.getLogger(PlacementSyncService.class);
+
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/PostSyncService.java:41]: Can use the `@Slf4j` annotation to create a logger for use named `log`, should be in the latest code from master.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java [hunk 10]
```diff
@@ -70,13 +72,16 @@
...
+
   private final Map<String, Function<Record, TraineeDetailsDto>> tableNameToMappingFunction;
 
   @Value("${service.trainee.url}")
   private String serviceUrl;
 
-  TcsSyncService(RestTemplate restTemplate, TraineeDetailsMapper mapper) {
+  TcsSyncService(RestTemplate restTemplate, TraineeDetailsMapper mapper, AmazonSQS amazonSQS) {
     this.restTemplate = restTemplate;
+    this.amazonSQS = amazonSQS;
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncService.java:84]: Looks like we can revert this class now the behaviour has been moved to the individual type services.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/TrustSyncService.java [hunk 11]
```diff
@@ -32,10 +35,13 @@
 @Service("reference-Trust")
 public class TrustSyncService implements SyncService {
 
+  private static final Logger LOG = LoggerFactory.getLogger(PlacementSyncService.class);
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/TrustSyncService.java:38]: Get latest from master, which includes `@Slf4j` annotation for logger.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/facade/PlacementEnricherFacadeTest.java [hunk 12]
```diff
@@ -107,7 +107,6 @@ void shouldEnrichFromPlacementWhenPostAndSameTrustsExist() {
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/facade/PlacementEnricherFacadeTest.java:107]: Revert the changes to this class, we should let the enricher take care of when/what to request so these verifications will be needed again.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncServiceTest.java [hunk 13]
```diff
@@ -114,7 +130,14 @@ void shouldNotFindRecordByIdPostWhenNotExists() {
...
-    assertThrows(UnsupportedOperationException.class, () -> service.request(ID));
+  void shouldEnrichARequestPost() throws JsonProcessingException {
+    when(postSyncService.findById(ID)).thenReturn(Optional.<Post>empty());
+
+    Map<String, String> recordData = record.getData();
+    recordData.put("postId", ID);
+
+    service.enrichOrRequestPost(record);
+
+    verify(postSyncService).request(ID);
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncServiceTest.java:141]: Need to implement and test the `request` method, other changes should be reverted.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncServiceTest.java [hunk 14]
```diff
@@ -60,6 +63,8 @@
 
   private static final String REQUIRED_ROLE = "DR in Training";
 
+  private AmazonSQS amazonSQS;
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncServiceTest.java:66]: Incorrect casing.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncServiceTest.java [hunk 15]
```diff
@@ -77,7 +82,8 @@ void setUp() {
     ReflectionUtils.setField(field, mapper, new TraineeDetailsUtil());
 
     restTemplate = mock(RestTemplate.class);
-    service = new TcsSyncService(restTemplate, mapper);
+    amazonSQS = AmazonSQSClientBuilder.defaultClient();
+    service = new TcsSyncService(restTemplate, mapper, amazonSQS);
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/service/TcsSyncServiceTest.java:86]: This can probably be reverted along with the TCS service

**Review by Judge40 — CHANGES_REQUESTED** [review]
Missing a few commits that merged yesterday which will help you tidy up a little.

`PlacementSyncService` has been made to do too much, the enricher takes care of requesting missing data based on mongo save events, so we don't need to do it again.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+   */
+  public DataRequestService(AmazonSQS amazonSqs,
+                            ObjectMapper objectMapper,
+                            @Value("${application.aws.sqs.queueUrl}") String queueUrl) {
+    this.amazonSqs = amazonSqs;
+    this.objectMapper = objectMapper;
+    this.queueUrl = queueUrl;
+  }
+
+  public void sendMessage(String tableName, String id) throws JsonProcessingException {
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java:34]: It would be nice to change this to match the change in class name. For example `sendMessage()` -> `sendRequest()`?

File: src/main/java/uk/nhs/hee/tis/trainee/sync/config/AmazonSqsConfig.java [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+    return new AWSStaticCredentialsProvider(basic);
+  }
+
+  @Bean
+  public AmazonSQS amazonSqs() {
+    AmazonSQSClientBuilder amazonSQSClientBuilder = AmazonSQSClientBuilder.standard();
+    amazonSQSClientBuilder.setRegion(region);
+    amazonSQSClientBuilder.setCredentials(credentialsProvider());
+    amazonSQSClientBuilder.setClientConfiguration(new ClientConfiguration());
+    return amazonSQSClientBuilder.build();
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/config/AmazonSqsConfig.java:37]: We should use the `AmazonSQSClientBuild.defaultClient()` and use the service's task role to provide credentials as env vars instead of handling them ourselves.

`tis-common-upload` takes a similar approach, I'm making the same changes in `tis-sync` at the moment. I can help you out with this change.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java [hunk 18]
```diff
@@ -0,0 +1,49 @@
+package uk.nhs.hee.tis.trainee.sync.service;
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestService.java:1]: Missing copyright notice.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/config/AmazonSqsConfig.java [hunk 19]
```diff
@@ -0,0 +1,40 @@
+package uk.nhs.hee.tis.trainee.sync.config;
```
Comment (Judge40) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/config/AmazonSqsConfig.java:1]: Missing copyright notice.

File: src/main/resources/application.yml [hunk 20]
```diff
@@ -17,3 +17,15 @@ spring:
...
+
+application:
+  aws:
+    access:
+      key:
+        id: ${AWS_ACCESS_KEY_ID}
+    secret:
+      access:
+        key: ${AWS_SECRET_ACCESS_KEY}
+    region: ${REGION:eu-west-2}
```
Comment (Judge40) [inline src/main/resources/application.yml:29]: We won't need to provide these values if we use `defaultClient()`

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestServiceTest.java [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+import com.amazonaws.services.sqs.model.SendMessageRequest;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.junit.jupiter.MockitoExtension;
+import static org.mockito.Mockito.*;
+
+@ExtendWith(MockitoExtension.class)
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestServiceTest.java:13]: Don't need to specify this if we're creating mocks manually and not using annotations.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestServiceTest.java [hunk 22]
```diff
@@ -0,0 +1,37 @@
+package uk.nhs.hee.tis.trainee.sync.service;
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/service/DataRequestServiceTest.java:1]: Missing copyright notice.

File: src/test/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncServiceTest.java [hunk 23]
```diff
@@ -114,7 +123,9 @@ void shouldNotFindRecordByIdPostWhenNotExists() {
   }
 
   @Test
-  void shouldSendRetrievalRequest() {
-    assertThrows(UnsupportedOperationException.class, () -> service.request(ID));
+  void shouldSendARequestForPlacementToTisSync() throws JsonProcessingException {
```
Comment (Judge40) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncServiceTest.java:126]: Not a major issue, but the naming suggests we're able to test where the message goes. We only verify that a request is sent.

File: src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java [hunk 24]
```diff
@@ -54,12 +58,19 @@ public void syncRecord(Record record) {
...
+
   public Set<Placement> findByPostId(String postId) {
     return repository.findByPostId(postId);
   }
 
   public void request(String id) {
-    // TODO: Implement.
-    throw new UnsupportedOperationException();
+    log.info("Sending request for Post [{}]", id);
+    String table = "Post";
```
Comment (JosephKelly) [inline src/main/java/uk/nhs/hee/tis/trainee/sync/service/PlacementSyncService.java:69]: I think table should be `"Placement"`.  I'd refactor this to put the literal on line 71 `dataRequestService.sendMessage("Placement", id);`.

File: src/test/resources/application.yml [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+    url: http://${service.trainee.host}:${service.trainee.port}/trainee
+
+spring:
+  data:
+    mongodb:
+      uri: mongodb://${DB_USER:admin}:${DB_PASSWORD:pwd}@${DB_HOST:localhost}:${DB_PORT:27017}/${DB_NAME:sync}?authSource=${AUTH_SOURCE:admin}
+
+application:
+  aws:
+    access:
```
Comment (Judge40) [inline src/test/resources/application.yml:23]: We don't need these values if we use `defaultClient()`.

**Review by Judge40 — DISMISSED** [review]

**Review by JosephKelly — CHANGES_REQUESTED** [review]

File: src/test/java/uk/nhs/hee/tis/trainee/sync/TisTraineeSyncApplicationTest.java [hunk 26]
```diff
@@ -21,12 +21,21 @@
...
 
 @SpringBootTest
 class TisTraineeSyncApplicationTest {
 
+  @MockBean
+  AmazonSQS amazonSqs;
+
+  @MockBean
+  AmazonSqsConfig amazonSqsConfig;
+
```
Comment (JosephKelly) [inline src/test/java/uk/nhs/hee/tis/trainee/sync/TisTraineeSyncApplicationTest.java:38]: Appears this is unused

**Review by JosephKelly — APPROVED** [review]
